src/nupic/regions/SPRegion.cpp

```diff
@@ -0,0 +1,1239 @@
+  case 't':
+    if (name == "topDownMode") {
+      return args_.topDownMode;
+    }
+    break;
+  }                                                         // end switch
+  return this->RegionImpl::getParameterUInt32(name, index); // default
+}
+
+Int32 SPRegion::getParameterInt32(const std::string &name, Int64 index) {
+  if (name == "seed") {
+    return args_.seed;
+  }
+  return this->RegionImpl::getParameterInt32(name, index); // default
+}
+
+Real32 SPRegion::getParameterReal32(const std::string &name, Int64 index) {
+  switch (name[0]) {
+  case 'b':
+    if (name == "boostStrength") {
+      if (sp_)
+        return sp_->getBoostStrength();
+      else
+        return args_.boostStrength;
+    }
+    break;
+  case 'l':
+    if (name == "localAreaDensity") {
+      if (sp_)
+        return sp_->getLocalAreaDensity();
+      else
+        return args_.localAreaDensity;
+    }
+    break;
+  case 'm':
+    if (name == "minPctOverlapDutyCycles") {
+      if (sp_)
+        return sp_->getMinPctOverlapDutyCycles();
+      else
+        return args_.minPctOverlapDutyCycles;
+    }
+    break;
+  case 'p':
+    if (name == "potentialPct") {
+      if (sp_)
+        return sp_->getPotentialPct();
+      else
+        return args_.potentialPct;
+    }
+    break;
+  case 's':
+    if (name == "synPermInactiveDec") {
+      if (sp_)
+        return sp_->getSynPermInactiveDec();
+      else
+        return args_.synPermInactiveDec;
+    }
+    if (name == "synPermActiveInc") {
+      if (sp_)
+        return sp_->getSynPermActiveInc();
+      else
+        return args_.synPermActiveInc;
+    }
+    if (name == "synPermConnected") {
+      if (sp_)
+        return sp_->getSynPermConnected();
+      else
+        return args_.synPermConnected;
+    }
+    break;
+  }
+  return this->RegionImpl::getParameterReal32(name, index); // default
+}
+
+bool SPRegion::getParameterBool(const std::string &name, Int64 index) {
+  if (name == "globalInhibition") {
+    if (sp_)
+      return sp_->getGlobalInhibition();
+    else
+      return args_.globalInhibition;
+  }
+  if (name == "wrapAround") {
+    if (sp_)
+      return sp_->getWrapAround();
+    else
+      return args_.wrapAround;
+  }
+  return this->RegionImpl::getParameterBool(name, index); // default
+}
+
+// copy the contents of the requested array into the caller's array.
+// Allocate the buffer if one is not provided.  Convert data types if needed.
+void SPRegion::getParameterArray(const std::string &name, Int64 index, Array &array) {
+  if (name == "spatialPoolerInput") {
+    getInput("bottomUpIn")->getData().convertInto(array);
+  } else if (name == "spatialPoolerOutput") {
+    getOutput("bottomUpOut")->getData().convertInto(array);
+  } else if (name == "spInputNonZeros") {
+    if (!nzInputValid_) {
+      const Array &incoming = getInput("bottomUpIn")->getData(); //transform incoming to sparse format
+      const SDR nonZero = *incoming.getSDR();
+      nzInput_ = Array(nonZero);
```

Comment:
is this the correct way to replace Array::nonZero() ?

Reply:
What this is trying to do is if spOutputNonZeros parameter is requested, return the output in sparse format.  So... is the output buffer using SDR's for the SP algorithm yet?  If it is then just get the SDR that is returned from the SP algorithm, and call getFlatSparse( ) which will return a vector that you can return.  In order to get the plugin to expect Arrays with an SDR and to output Arrays containing SDR's you need to tell NetworkAPI that you are expecting them by setting NTA_BasicType_SDR as the type in the Spec for the bottomUpOut output and bottomUpIn input.  The link will then create them as SDR type Arrays.

If you are not using SDR's in the algorithm yet, then you will need to create an SDR, populate it with the bottomUpOut buffer contents, then you can call the getFlatSparse( ) to get the FlatSparse buffer to return..

Actually, if we used an SDR in the Array for the bottomUpOut output, the spOutputNonZeros parameter is obsolete because the caller can get it from the output.

Reply:
Hmmm, the spOutputNonZeros parameter is expecting to return an Array, not a vector.  If the bottomUpOut is  an NTA_BasicType_SDR type, then all you need to do is return the bottomUpOut Array.

Reply:
If you don't have an SDR type for bottomUpOut (i.e. the algorithm is not setup to handle them yet) then you can:
```
const Array& bottomUpOut = getOutput("bottomUpOut")->getData();
return bottomUpOut.get_as(NTA_BasicType_SDR);
```

Reply:
I'm lost here, when you have time, can you have a look, please?

Reply:
Try this:
```
// copy the contents of the requested array into the caller's array.
// Allocate the buffer if one is not provided.  Convert data types if needed.
// The returned Array will be type NTA_BasicType_SDR which contains both dense 
// and sparse formats.
void SPRegion::getParameterArray(const std::string &name, Int64 index, Array &array) {
  if (name == "spatialPoolerInput") {
      return getInput("bottomUpIn")->getData().get_as(NTA_BasicType_SDR);
  } else if (name == "spatialPoolerOutput") {
     return getOutput("bottomUpOut")->getData().get_as(NTA_BasicType_SDR);
  } else if (name == "spInputNonZeros") {
      return getInput("bottomUpIn")->getData().get_as(NTA_BasicType_SDR);
  } else if (name == "spOutputNonZeros") {
     return getOutput("bottomUpOut")->getData().get_as(NTA_BasicType_SDR);
  }
  else {
    this->RegionImpl::getParameterArray(name, index, array);
  }
}
```

Reply:
oops, its not reutnred. Its an argument.
```
// copy the contents of the requested array into the caller's array.
// Allocate the buffer if one is not provided.  Convert data types if needed.
// The returned Array will be type NTA_BasicType_SDR which contains both dense 
// and sparse formats.
void SPRegion::getParameterArray(const std::string &name, Int64 index, Array &array) {
  if (name == "spatialPoolerInput") {
      array = getInput("bottomUpIn")->getData().get_as(NTA_BasicType_SDR);
  } else if (name == "spatialPoolerOutput") {
     array = getOutput("bottomUpOut")->getData().get_as(NTA_BasicType_SDR);
  } else if (name == "spInputNonZeros") {
      array = getInput("bottomUpIn")->getData().get_as(NTA_BasicType_SDR);
  } else if (name == "spOutputNonZeros") {
     array = getOutput("bottomUpOut")->getData().get_as(NTA_BasicType_SDR);
  }
  else {
    this->RegionImpl::getParameterArray(name, index, array);
  }
}
```

Reply:
In the unit tests; SPRegionTest, line 330 to 336 is expecting NTA_BasicType_UInt32 so it will fail.  Changing it to NTA_BasicType_SDR will fix that but 
compareParameterArrays( ) is only going to compare the dense format.

So, since we ripped out the NTA_BasicType_Sparse, we don't have a way to represent the sparse format as anything other than an SDR.   I guess we could change the test to also compare the sparse content but that capability is already covered in the SDR tests.  So perhaps comparing the dense format is sufficient.